Fix and enhance support for different bazel metadata versions

abraemer opened the pull request:
I noticed that the conditions deciding between the two version for `.bzl` files are somewhat wrong. In Python `'a' and 'b' in some_string` is equal to `('a' and 'b') in some_string` and thus simplifies to `'b' in some_string`. In this instance, I think it shouldn't make a huge difference because fortunately the strings `upstream_address` and `vcs_commit_hash` still somewhat distinguish between versions.

I think just fixing the conditions makes them too restrictive (AFAIK there is no real standard for these METADATA.bzl files), so I combined both paths into one. This ensures that we always extract as much information as possible. 

I also added the option to use information from a `package_url` field which is important to me as we often have some `METADATA.bzl` files which are generated from Maven dependencies and thus the `name` is invalid by PURL specification. I wrote a somewhat realistic test case for this.

### Tasks

* [x] Reviewed [contribution guidelines](https://github.com/nexB/scancode-toolkit/blob/develop/CONTRIBUTING.rst)
* [x] PR is descriptively titled 📑 and links the original issue above 🔗
* [x] Tests pass -- look for a green checkbox ✔️ a few minutes after opening your PR
  Run [tests](https://scancode-toolkit.readthedocs.io/en/latest/contribute/contrib_dev.html#running-tests) locally to check for errors. 
* [x] Commits are in uniquely-named feature branch and has no merge conflicts 📁

Closes #4196

abraemer committed:
Fix conditions for bzl package versions

Signed-off-by: Adrian Braemer <adrian.braemer@tngtech.com>

abraemer committed:
add myself to AUTHORS

Signed-off-by: Adrian Braemer <adrian.braemer@tngtech.com>

abraemer committed:
combine paths for different bzl versions

* by combining the paths we ensure to extract maximal information
* I also added the possibility to extract information from a 'package_url' field

Signed-off-by: Adrian Braemer <adrian.braemer@tngtech.com>

abraemer commented:
Test failures seem unrelated to my changes

AyanSinhaMahapatra commented on tests/packagedcode/data/build/metadatabzl/with-package-url/METADATA.bzl:
```diff
@@ -0,0 +1,12 @@
+    "name": "androidx.compose.animation:animation",
+    "upstream_address": "https://developer.android.com/jetpack/androidx/releases/compose-animation#0.0.1",
+    "version": "0.0.1",
+    "package_url" : "pkg:maven/androidx.compose.animation/animation@0.0.1"
```
Could you maybe link to some examples of this type of manifests with `package_url` fields and maybe add one of those as tests, it's best to use real world examples probably. Or if you got this from some real example, you can also link that file here.

abraemer replied:
Unfortunately, I cannot link you one of our internal files. However my example file is very close to an actual file (the version number is wrong but that's all).

AyanSinhaMahapatra replied:
> Unfortunately, I cannot link you one of our internal files.

No worries on that obviously, but can you find some other real examples with `package_url` anywhere on github/elsewhere, or is this relatively new?

abraemer replied:
I searched on GitHub but did not find any other files that use that field but also I did not find many examples of bzl files at all. So perhaps it is not very common (yet?).
As this doesn't interfere with other things, would you be okay with merging this anyways? After all it just adds another possible data source for bzl-files that has the potential to be useful for others :)

AyanSinhaMahapatra commented on src/packagedcode/build.py:
```diff
@@ -374,54 +375,27 @@ def parse(cls, location, package_only=True):
+            download_url=metadata_fields.get('download_url'),
+            vcs_url=metadata_fields.get('vcs_url'),
+            sha1=metadata_fields.get('download_archive_sha1'),
+            # TODO: Store 'upstream_hash` somewhere
```
Could you add a line here to store `upstream_hash` in the `extra_data` mapping for this PackageData?

abraemer replied:
Done

AyanSinhaMahapatra commented on src/packagedcode/build.py:
```diff
@@ -374,54 +375,27 @@ def parse(cls, location, package_only=True):
+            # TODO: Store 'upstream_hash` somewhere
+        )
+        if 'vcs_commit_hash' in metadata_fields:
+            package_data["extra_data"] = dict(vcs_commit_hash=metadata_fields['vcs_commit_hash'])
```
Wouldn't it be better to do a get() rather than checking if this exists in the mapping?

abraemer replied:
When including the `upstream_hash`, I rewrote this part. Now we always create the `extra_data` dict with all keys and I use `.get` to extract the values.

AyanSinhaMahapatra reviewed (changes requested):
Thanks a lot @abraemer , a couple nits for your consideration. Looking good otherwise.

abraemer committed:
fix: minor changes to address review comments

* always create extra_data dictionary
* use get to extract information from metadatafields instead of branches
* also extract upstream_hash

Signed-off-by: Adrian Braemer <adrian.braemer@tngtech.com>

abraemer commented:
friendly ping since it has been 2 weeks now and I think this ready :)  What do you think @AyanSinhaMahapatra?

AyanSinhaMahapatra commented on src/licensedcode/juliacode.jl:
```diff
@@ -0,0 +1,175 @@
+set_high_intersection_filter(set::BitSet, cutoff) = filter(<=(cutoff), set)
+set_high_intersection_filter(set::MSet, cutoff) = filter(pair -> pair.first<=cutoff, set)
+
+function compare_token_sets(qset, iset, len_legalese, min_matched_length_high, min_matched_length; minimum_containment=0, high_resemblance_threshold=0.8)
```
This file and the code though looks quite interesting, doesn't seem to be related to licensedcode at all and is in julia which we don't use  :P maybe you commited this by mistake? Curious about what you're trying to do here, some kind of post-processing on scancode license detection results?

abraemer replied:
Oh wow how did that end up here? :sweat_smile: 
This from an experiment where I rewrote a small, self-contained part in Julia to see whether that speeds things up (it did but only in single-threaded mode so far).
Anyhow, good catch - I removed the file again.

AyanSinhaMahapatra commented on src/packagedcode/build.py:
```diff
@@ -374,54 +375,28 @@ def parse(cls, location, package_only=True):
+            download_url=metadata_fields.get('download_url'),
+            vcs_url=metadata_fields.get('vcs_url'),
+            sha1=metadata_fields.get('download_archive_sha1'),
+            # TODO: Store 'upstream_hash` somewhere
```
Coudl you also remove the `TODO` line here since you've addressed that? Thanks

AyanSinhaMahapatra commented on src/packagedcode/build.py:
```diff
@@ -374,54 +375,28 @@ def parse(cls, location, package_only=True):
+        # then create a package of that package type from the metadata info
+        package_data = dict(
+            datasource_id=cls.datasource_id,
+            type=metadata_fields.get('upstream_type', metadata_fields.get('package_type', cls.default_package_type)),
```
Very minor nit: These 2 levels of `.get()` and defaults on fallback are not particularly readable even though it's correct, what do you think about changing this to a if-else block with something like:
```
if 'upstream_type' in metadata_fields:
   type = metadata_fields.get('upstream_type')
elif 'package_type' in metadata_fields:
   type = metadata_fields.get('package_type')
else:
   type = cls.default_package_type
```
Or something else which is more readable like this? Same for `extracted_license_statement` and  `homepage_url` below.
Usually this is why we have variables with more complicated extraction defined before we create the `package_data = dict()` mapping, and only keep simple `.get()` operations with defaults there.

abraemer replied:
I rewrote the nested `.get`s to explicit `if-else` statements. In addition, I also refactored the `extra_data` to follow the other fields.

AyanSinhaMahapatra reviewed (changes requested):
Thanks for the changes @abraemer, this is almost ready with a couple small changes, apologies for the late review :sweat_smile:

abraemer commented:
@AyanSinhaMahapatra friendly ping :)

AyanSinhaMahapatra commented on src/packagedcode/build.py:
```diff
@@ -374,54 +375,49 @@ def parse(cls, location, package_only=True):
+            homepage_url = metadata_fields.get('homepage_url')
+        
+        extra_data = dict(
+            vcs_commit_hash=metadata_fields.get('vcs_commit_hash'),
```
These should be added only if they are present, optionally. Otherwise we have these keys with empty values. See the only remaining test failures because of this: https://dev.azure.com/nexB/scancode-toolkit/_build/results?buildId=15913&view=logs&j=1412dbfa-421a-5a40-8ddf-7073fc46aa19&t=13c90161-2109-5957-f563-a96fefc7d869

AyanSinhaMahapatra reviewed (commented):
Thanks++ @abraemer, last nit for your considerations. Ready to merge otherwise. Thanks for your patience.  

Could you also merge from develop again, should get rid of the flaky test failures.

abraemer commented:
@AyanSinhaMahapatra last nit is fixed and branch is up-to-date with develop. Let's see what CI says :)

abraemer committed:
fix: Adapt test to include new upstream hash

As we now store the upstream hash in the field `extra_data`, we need to make the test aware of that.

Signed-off-by: Adrian Braemer <adrian.braemer@tngtech.com>

AyanSinhaMahapatra commented:
Thanks++ @abraemer, all green and merging!
